### ISSUE TITLE
Disabling Drag & Drop

### ISSUE THREAD

**Issue by VijayanRamachandran** [MicrosoftEdge/WebView2Feedback#278]
**Usecase and Current Behaviour:**

1. In a WPF application i hosted a single instance of WebView2 as below, and upon application startup WebView2 navigated to appropriate URL as expected.
 
![image](https://user-images.githubusercontent.com/25306014/84739734-d54b7180-afc9-11ea-9e5d-f7e1d4814cc9.png)

2. When i drag and drop any file to WebView2 Control then a new Edge window popped up with content of dropped file as seen in below GIF. 

![WebView2-Drag Drop](https://user-images.githubusercontent.com/25306014/84739932-1cd1fd80-afca-11ea-9755-bfa5a7212074.gif)

**Expectation:** We want to disable drag and drop to WebView2 control. ( ie if any type of file is dropped on top of WebView2, it should not pop up a new Edge window with its content. )

I tried by setting attribute AllowDrop="False" to the WebView2 control as below. But still above explained drag and drop behavior is happening.

![image](https://user-images.githubusercontent.com/25306014/84740437-024c5400-afcb-11ea-92fb-8251604c9018.png)

**Version Information**
Microsoft.Web.WebView2  => 0.9.515-prerelease

Can you please help us about how to disable drag and drop functionality to WebView2 ? 
( If there is some reference implementation to disable drag and drop it would be helpful. )

[AB#27010207](https://microsoft.visualstudio.com/90b2a23c-cab8-4e7c-90e7-a977f32c1f5d/_workitems/edit/27010207)

**Comment by Sulquendi** [comment]
Have a look at the `AllowDrop` propertiy (setting it to `false`)

**Comment by pagoe-msft** [comment]
@VijayanRamachandran,

Great feature request! This is something we are currently working on right now. Currently, the WebView2 control mimics the behavior of the browser - when a file is dragged/dropped into the WebView2 the control navigates to the file. Would you like to disable drag and drop completely? Do you have a need to drag and drop files into parts of your UI?

**Comment by VijayanRamachandran** [comment]
@pagoe-msft,

**Thanks for quick feedback...**
Nice to hear that already disabling Drag & Drop in WebView2 is in progress. 

We have both the requirements based on application. ie.. Some scenario we need to disable drag and drop completely and in some scenario we have to disable parts of UI.

**Comment by liminzhu** [comment]
@pagoe-msft can navigationStarting serve as a workaround for now?

**Comment by VijayanRamachandran** [comment]
@liminzhu , @pagoe-msft 

I tried same by registering for NavigationStarting event in WPF host end to try workaround, but we are not getting control in NavigationStarting event handler / or in drag and drop case NavigationStarting is not fired from WebView2 host.


![image](https://user-images.githubusercontent.com/25306014/85006857-4e88c700-b178-11ea-8deb-826ba6b9f861.png)

**Comment by pagoe-msft** [comment]
@VijayanRamachandran 

One possible work around for now can be to disable drag and drop via javascript:

await webView.CoreWebView2.AddScriptToExecuteOnDocumentCreatedAsync("window.addEventListener('dragover',function(e){e.preventDefault();},false);");

await webView.CoreWebView2.AddScriptToExecuteOnDocumentCreatedAsync("window.addEventListener('drop',function(e){e.preventDefault();}, false);");

Does this work for you?

**Comment by VijayanRamachandran** [comment]
@pagoe-msft,

As you mentioned i tried using above javascript as below, but still the same behaviour. ( ie., When i drag and drop any file to WebView2 Control then a new Edge window popped up with content of dropped file )

![image](https://user-images.githubusercontent.com/25306014/85059718-badbe880-b1c1-11ea-876a-cd4ccc40a70d.png)


I tried to execute the same javascript using "ExecuteScriptAsync" API in "NavigationStarting" event as below then drag and drop is disabled as expected after execution of "NavigationStarting" event handler. ( but in timing, before execution of "NavigationStarting" event handler if we drop any file then new Edge window popped up with content of dropped file )

![image](https://user-images.githubusercontent.com/25306014/85061237-1f984280-b1c4-11ea-9400-dfe53e7fd643.png)

PS: Same is the output as above when we execute the javascript using "ExecuteScriptAsync" API in "NavigationCompleted" event also.

One more thing i noticed..
If i execute the same javascript with help of "AddScriptToExecuteOnDocumentCreatedAsync" API and without "await" keyword as below then once the webcontent is rendered drag and drop is disabled as expected. ( Bu before webcontent is rendered if we drop any file then new Edge window popped up with content of dropped file )

![image](https://user-images.githubusercontent.com/25306014/85062643-756dea00-b1c6-11ea-897c-af4de3f168af.png)

PS: In this case without "await" keyword we are getting below warning on compilation.
"warning CS4014: Because this call is not awaited, execution of the current method continues before the call is completed. Consider applying the 'await' operator to the result of the call."

Overall consistent behavior is happening even though we are using javascript !!!!!

**Comment by pagoe-msft** [comment]
Interesting!

Perhaps instead of NavigationStarting, could you see if ExecuteScripyAsync works with the New Window Requested Event](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/dotnet/0-9-538/microsoft-web-webview2-core-corewebview2#newwindowrequested) 

The await keyword is necessary for any asynchronous call to ensure its execution is completed.

I'm glad to know the JavaScript fix worked! In general WebView2 content should be treated like all web content which relies on JavaScript.

**Comment by douglas-jordan** [comment]
Please do not disable features that make the Webview2 control less of a browser, my code uses it both in a static and dynamic applications.

**Comment by VijayanRamachandran** [comment]
@pagoe-msft 

As you mentioned i tried with "NewWindowRequested" event and "ExecuteScripyAsync" API to execute javascript as in below code snippet, but it is **not** working as expected.
When i drag and drop any file to WebView2 Control then a new Edge window popped up with content of dropped file. Drag and Drop is **not** disabled in this case..

![image](https://user-images.githubusercontent.com/25306014/85067880-08ab1d80-b1cf-11ea-9ef8-08a3da7306b7.png)

**Comment by pagoe-msft** [comment]
@VijayanRamachandran - thanks for checking that out for us. 

@douglas-jordan - that is our intention! Have you explored dragging and dropping content into your WebView2 control?

**Comment by VijayanRamachandran** [comment]
@pagoe-msft 

Welcome. 

As you mentioned in our conversation early that, the work is in progress in your end, can you please update us any specific timeline regarding this ?

**Comment by santoshmetri** [comment]
Hi Microsoft,

As mentioned in this thread, when we drag and drop the HTML file, then a new Edge window popped up with content of dropped file.

Our requirement is to display the file in the same WebView container on which the file is dropped.

Do we have any solution for above requirement yet?

Regards,
Santosh

**Comment by hugoam** [comment]
I found a way to deactivate this behavior:
```
            webView.CoreWebView2.NewWindowRequested += (sender, args) => args.Handled = true;
```
However, there's the opposite behavior (dragging something from inside the webview to anywhere in Windows explorer creates a link) which I'm trying to deactivate and still did not find a way to do so

**Comment by beretboat** [comment]
I can see that [AllowDrop](https://docs.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.winforms.webview2.allowdrop?view=webview2-dotnet-1.0.721-prerelease&viewFallbackFrom=webview2-dotnet-1.0.664.37) is available the .Net version of 1.0.721 pre-release. But what's the API equivalence in win32?

**Comment by ShaunLoganOracle** [comment]
> I can see that [AllowDrop](https://docs.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.winforms.webview2.allowdrop?view=webview2-dotnet-1.0.721-prerelease&viewFallbackFrom=webview2-dotnet-1.0.664.37) is available the .Net version of 1.0.721 pre-release.

And, it is read-only, so it won't suffice for suppressing drag/drop into the WebView2.

**Comment by darbid** [comment]
> Please do not disable features that make the Webview2 control less of a browser, my code uses it both in a static and dynamic applications.

I am in full agreement of this. This may not be the popular opinion but such a control should leave it to the developer to decide what features and security features he turns on and off. Leave all those non-changeable security features in Edge for the greater protection of the public who use it as a web-browser. A little off topic.

With respect to this topic: I have a use case. A user can drag a file onto the browser to upload it. If the file is encrypted eg. smime email then the webpage just rejects it. If the various drag and drop features of a standard control fired first then I would have the opportunity to detect this and under the right conditions fix the encryption and most importantly then let the un-encrpyted file to be added to the webpage.

**Comment by ukandrewc** [comment]
> he turns on and off

What?? Woman can't turn features on and off - who do I complain to?

**Comment by darbid** [comment]
I am so sorry for my non-gender neutral writing skills. They rank about as low as my developer skills.  I will say sorry for the next time I do this now as it is bound to happen again.

Oh and if you find out who to complain to let me know.

**Comment by ukandrewc** [comment]
@darbid ;-)

**Comment by dqwork** [comment]
just an extra voice saying, I'd love to be able to disable drop (of drag/drop) just like the old WebBrowser control could

**Comment by darbid** [comment]
I am going to make a wild guess and say that drag and drop is a lot of work for the WebView2 team as it is embeded in one of the Chromium windows.

If all you want to do is disable drag and drop that is "easy" :-)

[Have a look at my answer here on Stack Overflow](https://stackoverflow.com/a/54724688/678412) the topic is CefSharp but the underlying engine and/or windows (meaning for both WebView2 WPF and WinForms etc) are identical.  In a nutshell sending a RevokeDragDrop to the handle of the Chrome windows (not WebView2 HwndHost window) will kill any future drops.

`<DllImport("ole32.dll")>
Friend Shared Function RevokeDragDrop(ByVal hwnd As IntPtr) As IntPtr
End Function`

Make sure you only look at child windows of your app. If the user has MSTeams and Google Chrome for example running then they will have identically named windows.

**Comment by ekalchev** [comment]
> I am going to make a wild guess and say that drag and drop is a lot of work for the WebView2 team as it is embeded in one of the Chromium windows.
> 
> If all you want to do is disable drag and drop that is "easy" :-)
> 
> [Have a look at my answer here on Stack Overflow](https://stackoverflow.com/a/54724688/678412) the topic is CefSharp but the underlying engine and/or windows (meaning for both WebView2 WPF and WinForms etc) are identical. In a nutshell sending a RevokeDragDrop to the handle of the Chrome windows (not WebView2 HwndHost window) will kill any future drops.
> 
> _CefSharp examples show how to drill down into the Chromium windows but it is normally to get another window class. In this case I use the following (note I revoke all windows I find along the way in the call back function but it appears that only Chrome_WidgetWin_0 is registered._
> 
> `<DllImport("ole32.dll")> Friend Shared Function RevokeDragDrop(ByVal hwnd As IntPtr) As IntPtr End Function`
> 
> Make sure you only look at child windows of your app. If the user has MSTeams and Google Chrome for example running then they will have identically named windows.
> 
> By the way - Not too long ago Chromium did not recognize MSOutlook drags/drops. Once MS decided to dump Edge and make Edge Chromium it is my guess that MS engineers started working on the open source project and fixed the bug. So there are MS engineers who have worked on this area of Chromium before.

We are talking about WPF, this has nothing to do with it. While I use the same approach with CefSharp Winforms, WPF drag&drop is different.

**Comment by darbid** [comment]
@ekalchev may I suggest you review the original expectation of the person.

> We want to disable drag and drop to WebView2 control

What I supplied is exactly and currently as of April 2021 the only way to do this. 

Further if you take a look the underlying windows and class names are similar or the same in name at least, but appear to be created in Webview2 a little different.

**Comment by ekalchev** [comment]
WebView2 WPF is not wrapper around Win32. Check your facts.

**Comment by darbid** [comment]
Here you go.  
> This control is effectively a wrapper around the WebView2 COM API. You can directly access the underlying ICoreWebView2 interface and all of its functionality by accessing the CoreWebView2 property. Some of the most common COM functionality is also accessible directly through wrapper methods/properties/events on the control.

[Source of quote WebView2 Class for WPF](https://docs.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.wpf.webview2?view=webview2-dotnet-1.0.774.44)

[And A litte bit more on WPF Win32 Hosting ](https://docs.microsoft.com/en-us/dotnet/desktop/wpf/advanced/wpf-and-win32-interoperation?view=netframeworkdesktop-4.8). You too can take the winforms one and do the same and call it a WPF control if you like.

>The questions very clearly states that he was asking about WPF not WinForms. 

Very true. However the underlying Chromium parts are identical and that is where (currently) drag and drop is handled so subclassing these parts can help in certain circumstances.

>RevokeDragDrop cannot be used with WPF WebView2

Also very true if you add "cannot be used with WPF WebView2's **own handle**". But I would like to think in solutions not just problems.

**Comment by ekalchev** [comment]
You are right

**Comment by feich-ms** [comment]
Hi all, we are actively working on this feature that disable external drag and drops into webview. Here is the API spec we drafted https://github.com/MicrosoftEdge/WebView2Feedback/blob/master/specs/APIReview_AllowExternalDrop.md. FYI. If there are no upcoming concerns, we will move to next stage to enable this API in experimental release soon.

**Comment by heapwolf** [comment]
It seems like I'm able to intercept a native file drop with `webview->add_NewWindowRequested` and setting `put_Handled(true)` without specifying the instance of `put_NewWindow`. But this only provides me with with one file url. Does `get_AllowExternalDrop` fix this? If not, how can I get a list of all the files that the user is trying to drop so I can cancel the event and process the file list before deciding what to do next.

**Comment by feich-ms** [comment]
> It seems like I'm able to intercept a native file drop with `webview->add_NewWindowRequested` and setting `put_Handled(true)` without specifying the instance of `put_NewWindow`. But this only provides me with with one file url. Does `get_AllowExternalDrop` fix this? If not, how can I get a list of all the files that the user is trying to drop so I can cancel the event and process the file list before deciding what to do next.

What you mentioned above that using NewWindwoRequested related API to prevent new window launching looks a little tricky. Instead AllowExternalDrop should be recommended option to prevent dropping files into webview. Setting AllowExternalDrop to false will prevent all files dragging and dropping into webview, but you are free to toggle this property in your host app for specific conditions and the setting will take effect immediately. For what you request to get the file info that is trying to drop into webview, I think that will be a new feature request, could you please create another issue that describes your detail requests?

**Comment by heapwolf** [comment]
Here's what I want. I want to register `IDropTarget` on my `hWnd` and I want that to have precedence over webview2. Right now, my drop events fire on the titlebar, that's great. But webview2 takes over when I drag into it. I've looked at `put_AllowExternalDrop` but it looks experimental and subject to change. Is there another way to do this without waiting on some new APIs or feature requests?

**Comment by feich-ms** [comment]
> 

Cc @champnic, maybe Jameson has insights about the user's requirements?

**Comment by feich-ms** [comment]
> Here's what I want. I want to register `IDropTarget` on my `hWnd` and I want that to have precedence over webview2. Right now, my drop events fire on the titlebar, that's great. But my drag leaving event fires when I mouse over the webview2 component. Why does Webview2 need its own drag and drop api? If I can get the `{x,y}` and the list of files the user is trying to drop, I can communicate that to the webview, and the webview can do nothing. Less complexity would be a win. I've looked at `put_AllowExternalDrop` but it looks experimental and subject to change. Also, its really not obvious how to use the api. Is there another way to do this without waiting on new APIs or feature requests?

Cc @champnic, maybe Jameson has insights about the user's requirements?

**Comment by heapwolf** [comment]
@feich-ms also, I saw [this](https://github.com/MicrosoftEdge/WebView2Feedback/issues/1480#issuecomment-884753889) comment by @cmakloufi. So it seems disabling drag-drop from an existing API is possible. However, it's not clear to me how, in C++, to acquire a reference to the "grandchild window" in the `ICoreWebView2CreateCoreWebView2ControllerCompletedHandler` callback so that I can then call `RevokeDragDrop(hWnd)` on it.

**Comment by Eilon** [comment]
Related: https://github.com/MicrosoftEdge/WebView2Feedback/issues/1833

**Comment by NgoVuBaGiang** [comment]
hello,
I am trying to disable drag and drop from inside webview in Windows Form application, but i could only disable drag and drop partially. Here what is in my code

`webView.CoreWebView2.DOMContentLoaded += CoreWebView2_DOMContentLoaded;`

and then in CoreWebView2_DOMContentLoaded

> ` await webView.CoreWebView2.ExecuteScriptAsync("window.addEventListener('dragstart', function(e){e.preventDefault()};,false);" +
>                         "window.addEventListener('drag', function(e){e.preventDefault()};,false);" +
>                         "window.addEventListener('dragover',function(e){e.preventDefault();},false);" +
>                        "window.addEventListener('drop',function(e){" +
>                           "e.preventDefault();" +
>                           "console.log(e.dataTransfer);" +
>                           "console.log(e.dataTransfer.files[0])" +
>                        "}, false);");`

When app runs and user attempts to drag a hyperlink, a "X" icon will display, implicating that drag and drop is not possible. However, if user keeps dragging the hyperlink and drop it to the address bar of a web browser, the url will be pasted. How can I disable drag and drop completely?

**Comment by ajtruckle** [comment]
I stopped reading this topic half way though. My application has a browser control as part of a dialog. As such it is not a standard browser. It is not expected to behave like a browser. Therefore, by design, no one should be able to drag anything on to the browser and get it to display in a popup window.

How do I switch drag n drop off without messing with new windows and JavaScript?

**Comment by NgoVuBaGiang** [comment]
> I stopped reading this topic half way though. My application has a browser control as part of a dialog. As such it is not a standard browser. It is not expected to behave like a browser. Therefore, by design, no one should be able to drag anything on to the browser and get it to display in a popup window.
> 
> How do I switch drag n drop off without messing with new windows and JavaScript?
> 
> Why has the experimental API not been released? I use Win32.

I was tasked to disable Drag and Drop in windows application. My solution so far
`private async void CoreWebView2_DOMContentLoaded(object sender, CoreWebView2DOMContentLoadedEventArgs e)
        {
            //NOV1HC - Ngo Giang (SX/BSV33-EA)
            //Disable drag and drop from inside WebView
            await webView.CoreWebView2.ExecuteScriptAsync("window.addEventListener('dragstart', function(e){e.preventDefault();},false);" +
                        "window.addEventListener('drag', function(e){e.preventDefault();},false);" +
                        "window.addEventListener('dragover',function(e){e.preventDefault();},false);" +
                       "window.addEventListener('drop',function(e){e.preventDefault();}, false);" +
                       "window.document.getElementsByTagName('a').setAttribute('draggable','false');");
}`

**Comment by ajtruckle** [comment]
@NgoVuBaGiang Thanks. I can give it a go. It would be nice to use the api though. @champnic Why is this stalled from being stable?

**Comment by champnic** [comment]
Hey @ajtruckle - This is available in stable SDK 1.0.1185.39+:
https://learn.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2controller4?view=webview2-1.0.1661.34&preserve-view=true

Going to close this issue now. Not sure why we missed closing it a long time ago - apologies about the confusion!

**Comment by ajtruckle** [comment]
Hi @champnic 
Awesome! I was able to implement with a minor adjustment to the official C++ example:
```
wil::com_ptr<ICoreWebView2Controller4> controller4 = nullptr;
CHECK_FAILURE(m_pImpl->m_webController->QueryInterface(IID_PPV_ARGS(&controller4)));
if (controller4)
{
	CHECK_FAILURE(controller4->put_AllowExternalDrop(FALSE));
}
```
### Side question ###
You refer to **SDK 1.0.1185.39+** but when when I did the SDK version is my software it shows as:

> WebView2 SDK Build: 1661.34

When I debug the `get_TargetCompatibleBrowserVersion` results is actually shows: 111.0.1161.34. Why is the online documentation referring to 1.0.x yet the API referring to 111.0.x?

**Comment by champnic** [comment]
Hey @ajtruckle - The "SDK 1.0.1185.39+" is referring to the version of the SDK nuget package, which includes the definitions or the APIs you are able to access. The "111.0.1661.34" refers to the version of the WebView2 Runtime which implements the functionality. It looks like your currently used SDK is version 1.0.1661.34, which supports this API as it's later than the 1.0.1185.39, and your runtime also supports the API because it's build version 1661 is greater than 1185 as well.

**Comment by ajtruckle** [comment]
@champnic I get you. My code is based on the officialsample.

> WebView2 Runtime Version: 111.0.1661.62
> WebView2 SDK Build: 1661.34

As for NuGet, it says 1.0.1661.34.
